### ISSUE TITLE
Class is not generic if it extends `type[X]`

### ISSUE THREAD

**Issue by sobolevn** [python/cpython#100246]
Code sample:

```python
>>> from typing import TypeVar
>>> X = TypeVar('X')

>>> class Subtype(type[X]): ...
... 
>>> Subtype[int]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: type 'Subtype' is not subscriptable
```

This is quite strange. To make it generic, you must have explicit `Generic` base class:

```python
>>> from typing import Generic
>>> class Subtype(type[X], Generic[X]): ...
... 
>>> Subtype[int]
__main__.Subtype[int]
```

It looks like a bug to me because of two reasons:
1. Because other stdlib generics work differently:
```python
>>> class Sublist(list[X]): ...
... 
>>> Sublist[int]
```

2. Because `typing.Type` works as it should:

```python
>>> from typing import Type
>>> class Subtype(Type[X]): ...
... 
>>> Subtype[int]
__main__.Subtype[int]
```

But, I think that `type` is special enough to be handled extra carefully:
- `class S(type): ...` is not generic and does not have `__class_getitem__`, has `type` in `mro`
- `class S(type[X]): ...` is generic and has `__class_getitem__`, has `type` in `mro`

I can work on a fix if others agree that this is a bug 😊

**Comment by AlexWaygood** [comment]
I agree that this is inconsistent with other builtin generics. But what's the use case for subclassing `type[T]`, rather than just `type`? I can't think of a situation where it would be useful. If fixing this requires complicating the implementation, we need to think about what the benefits are from fixing this.

**Comment by sobolevn** [comment]
I am working with generic metaclasses that are designed to work with one specific object only. In my case: django models. I use this metaclass for validation and DI.

I am not sure about other use-cases, though.

**Comment by sobolevn** [comment]
The root cause for this is that `_BaseGenericAlias.__mro_entries__` adds `typing.Generic` to base classes if the type looks like generic:

https://github.com/python/cpython/blob/5693f45b19de47f8cb2f08d3baf43e626d3fbdf3/Lib/typing.py#L1272-L1282

While https://github.com/python/cpython/blob/5693f45b19de47f8cb2f08d3baf43e626d3fbdf3/Objects/genericaliasobject.c#L660-L665 does not do it.

**Comment by sobolevn** [comment]
This mypy problem is unrelated to the initial problem with the fact that type is not generic.
But, yes - type checkers is another topic.

For example, the initial problem occurs when I use this code:

```python
def validate_user(
   instance: User,
   validator_type: SemanticMeta[User],  # fails in runtime on this line, because it is not generic
): ...
```

**Comment by AlexWaygood** [comment]
> This mypy problem is unrelated to the initial problem with the fact that type is not generic.

But I tried running mypy on it, because I wanted to understand how type checkers understood the code in your proposed use case!

I guess I don't understand what using `class Meta(type[T]): ...` gets you here, as opposed to using `class Meta(type, Generic[T]): ...`. I understand the use case for generic metaclasses, but I'm still not sure I understand how type checkers are meant to interpret a metaclass inheriting from a specialised version of `type`.

How do these two metaclasses differ, from a type checker's point of view (in a world where type checkers are free of bugs)? Mypy already understands the `Meta2` definition just fine.

```python
from typing import TypeVar, Generic

class Foo: ...
FooT = TypeVar("FooT", bound=Foo)

class Meta1(type[FooT]):
    attr: FooT

class Meta2(type, Generic[FooT]):
    attr: FooT
```

**Comment by carljm** [comment]
IMO those two should be equivalent, just like `class Meta1(Type[FooT])` is equivalent to `class Meta2(Type, Generic[FooT])`. It does seem like it's not a critical gap, given that explicitly using `Generic` is an option, but it would be more consistent and less verbose. And it doesn't seem like the fix would be super invasive (but of course if that initial impression is wrong, it may not be worth it.)

**Comment by AlexWaygood** [comment]
I'm still not sure what the use case is for inheriting from either `type[T]` or `typing.Type[T]`, as opposed to using multiple inheritance with `Generic`. The specialisation of `type` doesn't seem to me to be "meaningful" in the same way that specialising `list` is when you're inheriting from it. 

For example, the following two classes mean very different things to the type checker:

```python
from typing import Generic, TypeVar

T = TypeVar("T")

class Foo(list[T]):
    extra_attr: T
    def __init__(self, extra_attr: T) -> None:
        super().__init__()
        self.extra_attr = extra_attr

class Bar(list, Generic[T]):
    extra_attr: T
    def __init__(self, extra_attr: T) -> None:
        super().__init__()
        self.extra_attr = extra_attr
```

Both classes are generic around `T`, but in the first class, the `extra_attr` attribute has to be of the same type as the elements in the list. In the second class, the `extra_attr` attribute could be an entirely different type to the elements in the list.

But while I feel like I understand how specialising `list` affects the semantics when you're subclassing it, I _don't_ feel like I understand how specialising `type` affects the semantics when you're subclassing it.

Having said all that, however, I agree that it would be nice for there to be consistency between `type[T]` and `typing.Type[T]`, so I don't want to block this if it's easy to implement.

**Comment by sobolevn** [comment]
I had time to get back to this. After further research, I don't think that adding `typing.Generic` is a right thing to do here. In fact, I think it is a wrong thing to do.

Why? Because `SubtypeOfBuiltinType[int]` is expected to return `types.GenericAlias`, not `typing._GenericAlias`.

Short demo:
```python
>>> from typing import *
>>> X = TypeVar('X')
>>> class A(list[X]): ...
... 
>>> class B(Type[X]): ...
... 
>>> type(A[int])
<class 'types.GenericAlias'>
>>> type(B[int])
<class 'typing._GenericAlias'>
```

So, subclasses of `type[X]` should also return `types.GenericAlias`.

But, `type` does not have this method at all. It is special cased in `PyObject_GetItem` here: https://github.com/python/cpython/blob/52bc2e7b9d451821513a580a9b73c20cfdcf2b21/Objects/abstract.c#L177-L183

So, I think that there's not *easy* fix:
1. We cannot return `typing._GenericAlias` or add `typing.Generic`, because it would make `type` inconsistent with other in a more devastating way
2. We cannot add `__class_getitem__` to `type` directly, because it would also add a lot of complexity / side effects

Some ideas on what can actually be done:
1. Add some special `_TypeGeneric` type into `mro`. But, this is quite complex and I am not sure about any potential side-effects
2. Add some more special-cases: if it is a subtype of `type` and does not have `__class_getitem__`, then use the `return Py_GenericAlias(o, key);`
3. Document this and leave as-is

I think that `3.` is the best way, actually.
